spacy/tests/training/test_training.py

```diff
@@ -1112,3 +1115,39 @@ def test_retokenized_docs(doc):
         retokenizer.merge(doc1[0:2])
         retokenizer.merge(doc1[5:7])
     assert example.get_aligned("ORTH", as_string=True) == expected2
+
+
+def test_training_before_update(doc):
+    def before_update(nlp, args):
+        assert args["step"] == 0
+        assert args["epoch"] == 1
+
+        # Raise an error here as the rest of the loop
+        # will not run to completion due to uninitialized
+        # models.
+        raise ValueError("ran_before_update")
+
+    def generate_batch():
+        yield 1, [Example(doc, doc)]
+
+    nlp = spacy.blank("en", config={"training": {}})
```

Comment:
Ah, I mean putting the `training.before_update` setting here? I guess it doesn't matter if you call `train_before_improving` directly?

Reply:
Yeah, it shouldn't matter.